triedb/pathdb/states.go

```diff
@@ -43,9 +49,528 @@ func (c *counter) report(count metrics.Meter, size metrics.Meter) {
+	return last, nil
+}
+
+func (j *journal) reset() {
+	j.destructs = nil
+}
+
+func (j *journal) encode(w io.Writer) error {
+	return rlp.Encode(w, j.destructs)
+}
+
+func (j *journal) decode(r *rlp.Stream) error {
+	var dec [][]destruct
+	if err := r.Decode(&dec); err != nil {
+		return err
+	}
+	j.destructs = dec
+	return nil
+}
+
+// stateSet represents a collection of state modifications belonging to a
+// transition(a block execution) or several aggregated transitions.
+type stateSet struct {
+	// destructSet is a very special helper marker. If an account is marked as
+	// deleted, then it's recorded in this set. However, it's allowed that an
+	// account is included here but still available in other sets (e.g.,
+	// accountData and storageData). The reason is the diff layer includes all
+	// the changes in a *block*. It can happen that:
+	//
+	// - in the tx_1, account A is deleted
+	// - in the tx_2, account A is recreated
+	//
+	// But we still need this marker to indicate the "old" A is deleted, all
+	// data in other set belongs to the "new" A.
+	destructSet map[common.Hash]struct{}               // Keyed markers for deleted (and potentially) recreated accounts
+	accountData map[common.Hash][]byte                 // Keyed accounts for direct retrieval (nil is not expected)
+	storageData map[common.Hash]map[common.Hash][]byte // Keyed storage slots for direct retrieval. one per account (nil means deleted)
+	size        uint64                                 // Memory size of the state data (destructSet, accountData and storageData)
+
+	journal           *journal                      // Track the modifications to destructSet, used for reversal
+	accountListSorted []common.Hash                 // List of account for iteration. If it exists, it's sorted, otherwise it's nil
+	storageListSorted map[common.Hash][]common.Hash // List of storage slots for iterated retrievals, one per account. Any existing lists are sorted if non-nil
+	lock              sync.RWMutex                  // Lock for guarding the two lists above
+}
+
+// newStates constructs the state set with the provided data.
+func newStates(destructs map[common.Hash]struct{}, accounts map[common.Hash][]byte, storages map[common.Hash]map[common.Hash][]byte) *stateSet {
+	// Don't panic for the lazy callers, initialize the nil maps instead.
+	if destructs == nil {
+		destructs = make(map[common.Hash]struct{})
+	}
+	if accounts == nil {
+		accounts = make(map[common.Hash][]byte)
+	}
+	if storages == nil {
+		storages = make(map[common.Hash]map[common.Hash][]byte)
+	}
+	s := &stateSet{
+		destructSet:       destructs,
+		accountData:       accounts,
+		storageData:       storages,
+		journal:           &journal{},
+		storageListSorted: make(map[common.Hash][]common.Hash),
+	}
+	s.size = s.check()
+	return s
+}
+
+// account returns the account data associated with the specified address hash.
+func (s *stateSet) account(hash common.Hash) ([]byte, bool) {
```

Comment:
I guess you're more or less doing the same locking as already existing code. E.g. `merge` happens when the disklayer does `commit`, and that only happens when the `disklayer` writelock is held. You can probably ignore my comments about locking

Reply:
Yes, the stateSet is an internal structure, it can only be held in diffLayer, or the buffer inside of diskLayer.

The stateSet itself doesn't need lock (it's lock free). Whenever the stateSet is modified (bottom-most diffLayer is merged into buffer), the original diskLayer will be marked as stale first, to prevent following access to the original stateSet. Then the new stateSet after mutation will be linked with new diskLayer.

In another word, the associated stateSet can be regarded as immutable (read-only) if the corresponding diffLayer or diskLayer is "unstale".